core/src/main/java/io/questdb/log/LogAlertSocket.java

```diff
@@ -0,0 +1,325 @@
+
+    public LogAlertSocket(FilesFacade ff, String socketAddress) {
+        this(ff, socketAddress, IN_BUFFER_SIZE, OUT_BUFFER_SIZE);
+    }
+
+    public LogAlertSocket(FilesFacade ff, String socketAddress, int outBufferSize) {
+        this(ff, socketAddress, IN_BUFFER_SIZE, outBufferSize);
+    }
+
+    public LogAlertSocket(FilesFacade ff, String socketAddress, int inBufferSize, int outBufferSize) {
+        this.ff = ff;
+        this.rand = new Rnd(System.currentTimeMillis(), System.currentTimeMillis());
+        this.socketAddress = socketAddress;
+        parseSocketAddress();
+        this.inBufferSize = inBufferSize;
+        this.inBufferPtr = Unsafe.malloc(inBufferSize, MemoryTag.NATIVE_DEFAULT);
+        this.outBufferSize = outBufferSize;
+        this.outBufferPtr = Unsafe.malloc(outBufferSize, MemoryTag.NATIVE_DEFAULT);
+    }
+
+    public void connect() {
+        fdSocketAddress = Net.sockaddr(hosts[hostPortIdx], ports[hostPortIdx]);
+        fdSocket = Net.socketTcp(true);
+        System.out.printf("Connecting with: %s:%d%n", hosts[hostPortIdx], ports[hostPortIdx]);
+        if (fdSocket > -1) {
+            if (Net.connect(fdSocket, fdSocketAddress) != 0) {
+                System.out.println(" E could not connect");
+                freeSocket();
+            }
+        } else {
+            System.out.println(" E could not create TCP socket [errno=" + ff.errno() + "]");
+            freeSocket();
+        }
+    }
+
+    public void send(int len) {
+        send(len, 0, null);
+    }
+
+    public void send(int len, Consumer<String> ackReceiver) {
+        send(len, 0, ackReceiver);
+    }
+
+    @Override
+    public void close() {
+        if (fdSocket != -1) {
+            freeSocket();
+        }
+        if (outBufferPtr != 0) {
+            Unsafe.free(outBufferPtr, outBufferSize, MemoryTag.NATIVE_DEFAULT);
+            outBufferPtr = 0;
+        }
+        if (inBufferPtr != 0) {
+            Unsafe.free(inBufferPtr, IN_BUFFER_SIZE, MemoryTag.NATIVE_DEFAULT);
+            inBufferPtr = 0;
+        }
+    }
+
+    public long getOutBufferPtr() {
+        return outBufferPtr;
+    }
+
+    public int getOutBufferSize() {
+        return outBufferSize;
+    }
+
+    public long getInBufferPtr() {
+        return inBufferPtr;
+    }
+
+    public long getInBufferSize() {
+        return inBufferSize;
+    }
+
+    private void send(int len, int failOverLevel, Consumer<String> ackReceiver) {
+        if (fdSocket > 0) {
+            //
+            int remaining = len;
+            long p = outBufferPtr;
+            boolean sendFail = false;
+            while (remaining > 0) {
+                int n = Net.send(fdSocket, p, remaining);
+                if (n > 0) {
+                    remaining -= n;
+                    p += n;
+                } else {
+                    System.out.println("could not send [n=" + n + " [errno=" + ff.errno() + "]");
+                    sendFail = true;
```

Comment:
this is an infinite loop. When socket begins to return negative result the loop does not exit. This clearly needs to be tested. I would suggest to use `NetworkFacade`  instead of `Net` and test it